### PR TITLE
Handle rate limiting error when fetching

### PR DESCRIPTION
**Pull request by evilmarty**
If CloudWatch raises a ThrottlingException it means there's too many requests. This error bubbles up and causes the plugin to crash in Logstash which then re-starts it. Instead we simply return from `process_group` and try again later on. Thus losing track since last last fetch (if not persisted).

### CODE REVIEW

**Comment by evilmarty** [comment]
I notice you've implemented your own throttling handler. Will close this PR.

**Comment by lukewaite** [comment]
@evilmarty Yeah - that's my bad. I saw the email subject, didn't notice a PR, and went "oh crap, that's a regression", and went to fix it. Then went looking for the issue, and noticed it was a PR, but that you'd implemented it slightly differently.

Your method would switch immediately to the next log_group (if multiple), which could then be throttled again, but would benefit from the set `interval` as a backoff attempt. The benefit here would be more cleanly exiting without having to implement any special handling to do so if being throttled and attempting to pull in vast amounts of data.

My method backs off, but stays in the current group until complete, before moving on. It however doesn't exit cleanly if asked to do so.

I'm still not sure which is better.

In general I'm open to opinions on this one..

**Comment by evilmarty** [comment]
@lukewaite I've updated my PR with a slight change to my approach. Instead of handling the error at the point of call I have moved the rescue up further to the main loop. As you pointed out we simply wait during the interval pause and try again albeit a different log group to ensure none are neglected. What are your thoughts?

**Comment by lukewaite** [comment]
@evilmarty Sorry for the delay in getting back to you on this one. First pass looks good, but I'd like to test locally before merging.

**Comment by evilmarty** [comment]
@lukewaite no worries. I've been using this code for over a week without hiccups in both my staging and production environments. Let me know how you go.

**Comment by lukewaite** [comment]
@evilmarty I've reverted 12565715f7752de08a99e7eabc184ab87a5f62d3 in prep for merging this, as it's not in your fork, and I don't think it's necessary with these changes.

File: lib/logstash/inputs/cloudwatch_logs.rb [hunk 1]
```diff
@@ -110,16 +110,21 @@ def check_start_position_validity
...
 
     while !stop?
-      groups = find_log_groups
-
-      groups.each do |group|
-        @logger.debug("calling process_group on #{group}")
-        process_group(group)
-      end # groups.each
+      begin
+        groups = find_log_groups
```
Comment (lukewaite) [inline lib/logstash/inputs/cloudwatch_logs.rb]: I'm wondering if I should move this line out of your begin/rescue here and add back in the handling for find_log_groups that was part of 12565715f7752de08a99e7eabc184ab87a5f62d3. 

I think that if we got throttled fetching log groups here - we might get into a situation where we're always being throttled, and just stopping.

It's probably a very rare corner case... but might be better to handle it differently than how we handle an interruption while processing a group.

Comment (evilmarty) [inline lib/logstash/inputs/cloudwatch_logs.rb]: `find_log_groups` hits the API as well which can be throttled. If it's not handled then it'll simply error and cause the whole plugin to reload (potentially losing position etc). With this change throttling is always handled for all API calls.

**Comment by evilmarty** [comment]
@lukewaite Anything further holding up this PR?

**Comment by lukewaite** [comment]
No blockers, @evilmarty, other than making the time for it which I haven't. My apologies.

I'll merge in now and tag a patch release.

**Comment by lukewaite** [comment]
Tagged now and published to rubygems.
https://github.com/lukewaite/logstash-input-cloudwatch-logs/releases/tag/v1.0.1

**Pull request merged**